### DISCUSSION TITLE
Including stable-baselines3 and torch & cuda dependencies in `requirements`

### DISCUSSION THREAD

**Discussion by hagrid67** [flatland-association/flatland-rl#63] [General]
We have introduced `stable-baselines3` into the dependencies - see `pyproject.toml` and the generated `requirements.txt` and `requirements-dev.txt`.  This package requires `torch` which in turn requires various nvidia packages; this makes the virtual env several Gb in size.  Previously we have separated ML-related libraries so that Flatland itself is a relatively lightweight and ML-agnostic package.

Are we happy with this change or do we need to discuss further?

(edit title from reqs-dev to reqs because currently baselines3 appears in **reqs**)

**Comment by fnberta** [comment]
As @manuschn mentioned on slack, no new dependencies were introduced when we moved to `pyproject.toml`. What we did however is make all the existing ones explicit. Previously, there were a lot of dependencies not listed anywhere. I did not special case the `contrib` folder in any way though. Maybe it is something special and you want to handle dependencies differently? I'm also seeing just now that there is a `requirements_training.txt` in there. Not sure what role this should play. In that sense, thanks for opening the discussion. @manuschn @aiAdrian what are your thoughts on this?

**Reply by hagrid67** [reply]
Yeah not sure the best way to handle it in the CI structure but I'd be inclined to make another package(s) which includes flatland and additionally the heavyweight GPU/ML packages (which FL itself doesn't yet use).  So training, baselines, solutions etc.  

I think we had something like that in the past but I'm not familiar with the structure we used, other than contrib.

**Comment by hagrid67** [comment]
Currently the `stables-baselines3` appears in the main requirements; IMO it should at least be demoted to `reqs-dev`,  and should be in `reqs-training` if that is available.

I take your point that it was always a **latent** dependency; but running the pettingzoo baselines was a fairly niche activity; someone doing that could easily pip install `torch` etc; it did not need to be an explicit dependency which affected everyone trying to just install flatland.  

flatland itself is like one of the many small environments in gym / pettingzoo: it doesn't know anything about training etc; it's just a small program with the capability to generate a graphical / visual output.  The packages for this are relatively small (a few kb to a few Mb, numpy is 36Mb)

The flatland dev requirements are for developing flatland; this could be seen as just developing the environment.  This includes some fairly bulky things like jupyter (~20Mb across a few packages?)

Do we have the facility for separate `requirements-training`, as well as `requirements-dev`? (I got the impression there was just a single bit switch to enable "optional-dependencies" in `pyproject.toml`, so the answer would appear to be no.  But if we do, then we could push all the GPU / ML packages and the packages which need them into `reqs-training`

The GPU / ML libraries (eg torch ~1.3Gb, nvidia ~2.6Gb) are an order of magnitude or two bigger than packages like numpy (36Mb).

FYI Some of the larger package sizes from the virtual-env I'm using:
13M     networkx
13M     rapidfuzz
15M     cryptography
15M     pip
16M     jupyterlab
17M     wandb
18M     debugpy
18M     lxml
24M     botocore
34M     babel
35M     scipy.libs
36M     numpy
37M     numpy.libs
38M     matplotlib
42M     nbclassic
52M     sympy
61M     pandas
72M     cmake
76M     panel
85M     bokeh
86M     scipy
115M    pyarrow
162M    ray
185M    triton
1.3G    torch
2.6G    nvidia

(created using the command `du -sh * | sort -h` in the `site-packages` directory; this is from my general virtual env which may contain stuff not required by flatland)

**Comment by hagrid67** [comment]
Meh - I've moved `stable_baselines3` from the dependencies into the optional dependencies so that `torch` and `nvidia` are no longer core `requirements` but just appear in `reqs-dev`.

**Comment by aiAdrian** [comment]
Great

hagrid67 ***@***.***> schrieb am Mi., 6. Dez. 2023, 10:00:
